evm/src/arithmetic/addcc.rs

```diff
@@ -0,0 +1,384 @@
+    x: &[ExtensionTarget<D>],
+    y: &[ExtensionTarget<D>],
+    z: &[ExtensionTarget<D>],
+    cy: ExtensionTarget<D>,
+    is_two_row_op: bool,
+) {
+    debug_assert!(x.len() == N_LIMBS && y.len() == N_LIMBS && z.len() == N_LIMBS);
+
+    // Since `map` is lazy and the closure passed to it borrows
+    // `builder`, we can't then borrow builder again below in the call
+    // to `eval_ext_circuit_pol_eval_equal`. The solution is to force
+    // evaluation with `collect`.
+    let pol_sum = x
+        .iter()
+        .zip(y)
+        .map(|(&xi, &yi)| builder.add_extension(xi, yi))
+        .collect::<Vec<ExtensionTarget<D>>>();
+
+    let expected_cy = eval_ext_circuit_pol_eval_equal(
+        builder,
+        yield_constr,
+        filter,
+        pol_sum.into_iter(),
+        z.iter().copied(),
+        is_two_row_op,
+    );
+    let good_cy = builder.sub_extension(cy, expected_cy);
+    let filter = builder.mul_extension(filter, good_cy);
+    if is_two_row_op {
+        yield_constr.constraint_transition(builder, filter);
+    } else {
+        yield_constr.constraint(builder, filter);
+    }
+}
+
+pub fn eval_ext_circuit<F: RichField + Extendable<D>, const D: usize>(
+    builder: &mut plonky2::plonk::circuit_builder::CircuitBuilder<F, D>,
+    lv: &[ExtensionTarget<D>; NUM_ARITH_COLUMNS],
+    yield_constr: &mut RecursiveConstraintConsumer<F, D>,
+) {
+    let is_add = lv[IS_ADD];
+    let is_sub = lv[IS_SUB];
+    let is_lt = lv[IS_LT];
+    let is_gt = lv[IS_GT];
+
+    let x = &lv[GENERAL_REGISTER_0];
+    let y = &lv[GENERAL_REGISTER_1];
+    let z = &lv[GENERAL_REGISTER_2];
+    let cy = lv[GENERAL_REGISTER_BIT];
+
+    let op_filter = builder.add_many_extension([is_add, is_sub, is_lt, is_gt]);
+    eval_ext_circuit_check_is_one_bit(builder, yield_constr, op_filter, cy);
+
+    eval_ext_circuit_add_cc(builder, yield_constr, op_filter, x, y, z, cy, false);
+}
+
+#[cfg(test)]
+mod tests {
+    use plonky2::field::goldilocks_field::GoldilocksField;
+    use plonky2::field::types::{Field, Sample};
+    use rand::{Rng, SeedableRng};
+    use rand_chacha::ChaCha8Rng;
+
+    use super::*;
+    use crate::arithmetic::columns::NUM_ARITH_COLUMNS;
+    use crate::constraint_consumer::ConstraintConsumer;
+
+    // TODO: Should be able to refactor this test to apply to all operations.
+    #[test]
+    fn generate_eval_consistency_not_addcc() {
+        type F = GoldilocksField;
+
+        let mut rng = ChaCha8Rng::seed_from_u64(0x6feb51b7ec230f25);
+        let mut lv = [F::default(); NUM_ARITH_COLUMNS].map(|_| F::sample(&mut rng));
+
+        // if the operation filters are all zero, then the constraints
+        // should be met even if all values are
+        // garbage.
+        lv[IS_ADD] = F::ZERO;
+        lv[IS_SUB] = F::ZERO;
+        lv[IS_LT] = F::ZERO;
+        lv[IS_GT] = F::ZERO;
+
+        let mut constrant_consumer = ConstraintConsumer::new(
+            vec![GoldilocksField(2), GoldilocksField(3), GoldilocksField(5)],
```

Comment:
nit: These should probably be more random than 2,3,5 to be certain nothing funny is going on.

Reply:
I'm going to address this in a follow-up PR that does more thorough testing throughout.